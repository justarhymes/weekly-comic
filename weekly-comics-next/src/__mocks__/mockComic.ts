import type { Comic } from '@/app/types';

export const mockComic: Comic = {
  id: 999,
  metron_id: 123456,
  comicvine_id: 5345,
  gcd_id: 9879,
  slug: 'batman-2025-1',
  title: 'Batman #1',
  issue_number: '1',
  summary: 'Bruce Wayne returns to Gotham as a new enemy emerges.',
  release_date: '2025-05-01',
  price: 3.99,
  image: 'https://example.com/batman-1.jpg',
  blur_data_url: '',
  upc: '123456789012',
  distributor_sku: 'BATMAN202501',
  rating: 'T+',
  page_count: 32,
  api_source: 'metron',
  is_custom: false,
  creators: [
    { name: 'Tom King', roles: ['Writer'] },
    { name: 'Greg Capullo', roles: ['Artist'] },
  ],
  series: {
    name: 'Batman',
    slug: 'batman',
    type: 'Ongoing',
    genre: ['Superhero'],
    volume: 1,
    publisher: 'DC Comics',
    start_year: 2025,
  },
};
