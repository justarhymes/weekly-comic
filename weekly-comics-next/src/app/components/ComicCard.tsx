import Link from "next/link";
import Image from "next/image";
import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";
import { Comic } from "@/app/types";

interface ComicCardProps {
  comic: Comic;
}

export default function ComicCard({ comic }: ComicCardProps) {
  const hasData = comic && comic.title && comic.price !== null;

  return (
    <div className="border border-foreground/10 rounded-lg overflow-hidden shadow-sm">
      {hasData ? (
        <Link href={`/comics/${comic.id}`}>
          <div>
            {comic.image ? (
              <Image
                src={comic.image}
                alt={comic.title}
                width={300}
                height={450}
                className="w-full h-auto object-cover"
              />
            ) : (
              <Skeleton height={450} />
            )}
            <div className="p-2">
              <h3 className="font-semibold text-sm line-clamp-2">
                {comic.title}
              </h3>
              { comic.price ? (
                <p className="text-xs text-gray-500">${comic.price?.toFixed(2)}</p>
              ) : "" }
            </div>
          </div>
        </Link>
      ) : (
        <>
          <Skeleton height={450} />
          <div className="p-2">
            <Skeleton count={2} />
          </div>
        </>
      )}
    </div>
  );
}