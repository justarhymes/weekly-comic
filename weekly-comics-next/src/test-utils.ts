import { render } from '@testing-library/react';
import type { ReactElement } from 'react';

// In the future, wrap with any Providers or Router if needed
export function renderWithProps(ui: ReactElement) {
  return render(ui);
}