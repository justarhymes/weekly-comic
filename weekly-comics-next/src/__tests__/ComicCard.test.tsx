import { screen } from '@testing-library/react';
import ComicCard from '@/app/components/ComicCard';
import { mockComic } from '@/__mocks__/mockComic';
import { renderWithProps } from '@/test-utils';

// ðŸ‘‡ Mock framer-motion to inspect props passed to motion.div
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...rest }: any) => (
      <div data-motion-props={JSON.stringify(rest)}>{children}</div>
    ),
  },
}));

describe('ComicCard', () => {
  it('renders the comic title and links to the detail page', () => {
    renderWithProps(<ComicCard comic={mockComic} index={0} />);
    const link = screen.getByRole('link', { name: /batman #1/i });
    expect(link).toHaveAttribute('href', `/comics/${mockComic.slug}`);
    expect(link).toHaveAttribute('title', 'View details for Batman #1');
  });

  it('displays the comic price correctly', () => {
    renderWithProps(<ComicCard comic={mockComic} index={0} />);
    expect(screen.getByText('$3.99')).toBeInTheDocument();
  });

  it('displays the comic image with appropriate alt text', () => {
    renderWithProps(<ComicCard comic={mockComic} index={0} />);
    const image = screen.getByAltText(/cover of batman #1/i);
    expect(image).toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const { container } = renderWithProps(<ComicCard comic={mockComic} index={0} />);
    expect(container).toMatchSnapshot();
  });

  it('applies animation delay based on index', () => {
    const index = 4;
    const { container } = renderWithProps(<ComicCard comic={mockComic} index={index} />);
    const motionDiv = container.querySelector('[data-motion-props]');
    expect(motionDiv).toBeInTheDocument();

    const props = JSON.parse(motionDiv!.getAttribute('data-motion-props') || '{}');
    expect(props.transition.delay).toBeCloseTo(index * 0.05);
  });
});
