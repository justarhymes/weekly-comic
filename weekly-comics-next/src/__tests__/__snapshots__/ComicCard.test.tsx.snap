// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicCard matches the snapshot 1`] = `
<div>
  <div
    data-motion-props="{"initial":{"opacity":0,"y":20},"animate":{"opacity":1,"y":0},"transition":{"duration":0.4,"ease":"easeOut","delay":0}}"
  >
    <a
      aria-describedby="comic-price-999"
      aria-labelledby="comic-999"
      class="focus:outline-none focus-visible:ring-4 focus-visible:ring-rose-500 block"
      href="/comics/batman-2025-1"
      title="View details for Batman #1"
    >
      <article
        class="group transition-transform duration-200 hover:scale-[1.03] active:scale-[0.98]"
      >
        <img
          alt="Cover of Batman #1"
          class="w-full h-auto object-cover transition duration-200 group-hover:opacity-90"
          data-nimg="1"
          decoding="async"
          height="450"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fexample.com%2Fbatman-1.jpg&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fbatman-1.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Fbatman-1.jpg&w=640&q=75 2x"
          style="color: transparent;"
          width="300"
        />
        <div
          class="bg-rose-500 p-2 transition-colors duration-200 group-hover:bg-rose-400"
        >
          <h3
            class="font-semibold text-sm line-clamp-1 text-white group-hover:text-rose-950 transition-colors duration-200"
            id="comic-999"
          >
            Batman #1
          </h3>
          <p
            class="text-rose-950 font-bold"
            id="comic-price-999"
          >
            $3.99
          </p>
        </div>
      </article>
    </a>
  </div>
</div>
`;
